#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)
a = as.integer(args[1])

library(caret)
library(Seurat)
library(org.Mm.eg.db)
library(GO.db)
library(AnnotationDbi)
library(dplyr)

a = as.integer(args[1])
b = as.integer(args[2])

OSN = readRDS("OSN_10.rds")
Go_ID = readRDS("Go_ID_BP.rds") # or Go_ID = readRDS("Go_ID_MF.rds") 
dirname = gsub(":","_",Go_ID[a])
dir.create(paste(dirname),showWarnings = F)
all_genes = unique(AnnotationDbi::select(org.Mm.eg.db, keys=Go_ID[a], columns=c("SYMBOL"), keytype="GO")[,"SYMBOL"])
all_genes = all_genes[all_genes%in%rownames(OSN)]
#for the genes related to bulk RNAseq:
#AG_genes = grep("Eph|Nrp|Kirrel|Efn|Pcdh|Sema|Plxn|Robo|Slit|Ntn|Wnt", rownames(OSN_10_rm), value = T)
#AG_genes = grep("Pcdha|Pcdhb|Pcdhg", AG_genes, value = T, invert = T)
genes = sample(all_genes, 150)


###generate train and test data
ORs = as.character(unique(OSN$OR_identity))
train.index <- createDataPartition(OSN$OR_identity, p = 0.75, list = FALSE)
test_cells = Cells(OSN)[-train.index]
train_cells = Cells(OSN)[train.index]
train_data = subset(OSN, cells = train_cells)
test_data = subset(OSN, cells = test_cells)
train_metadata = train_data@meta.data

#take min number of cells
ncells = min(table(train_data$OR_identity))

# split each OSN group of min number
chunks <- function(x,n) split(x, cut(seq_along(x), n, labels = FALSE))
ncells_list = list()
for (i in 1:length(ORs)) {
  sub =  train_metadata[train_metadata$OR_identity==ORs[i],]
  ncells_list[[i]] = chunks(sample(rownames(sub)), ncells)
}
# average the normalized transcriptome into avergae cells
train_metadata$group = NA
for (i in 1:ncells) {
  cell = unlist(lapply(ncells_list, `[[`, i))
  train_metadata$group[rownames(train_metadata)%in%cell]=i
}
train_metadata$ave_cell = paste0(train_metadata$OR_identity,"_",train_metadata$group)

train_data@meta.data = train_metadata
Idents(train_data) = train_data$ave_cell
train_data_ave = AverageExpression(train_data, return.seurat = T)
train_data_ave = FindVariableFeatures(train_data_ave)
train_data_ave$OR_identity = gsub("_.*","",colnames(train_data_ave))


train_data_ave_norm = train_data_ave@assays$RNA@data[genes,]
train_data_ave_norm = train_data_ave_norm[ - as.numeric(which(apply(train_data_ave_norm, 1, var) == 0)),]
genes = rownames(train_data_ave_norm)
train_data_norm = train_data@assays$RNA@data[genes,]

prin_comp <- prcomp(t(train_data_ave_norm), center = TRUE,scale. = TRUE)
# project all the train data onto the PCA space generated by the averaged train cells
projected_train_data = scale(t(as.matrix(train_data_norm)), prin_comp$center, prin_comp$scale) %*% prin_comp$rotation 
all(rownames(projected_train_data)==colnames(train_data))
projected_train_data = as.data.frame(projected_train_data)
projected_train_data$OR = train_data$OR_identity
saveRDS(prin_comp, paste0(dirname,"/pca_",b,".rds"))
write.csv(projected_train_data, paste0(dirname,"/train_",b,".csv"))

# project all the test data onto the PCA space generated by the averaged train cells
test_data_norm = test_data@assays$RNA@data[genes,]
projected_test_data = scale(t(as.matrix(test_data_norm)), prin_comp$center, prin_comp$scale) %*% prin_comp$rotation 
all(rownames(projected_test_data)==colnames(test_data))
projected_test_data = as.data.frame(projected_test_data)
projected_test_data$OR = test_data$OR_identity
write.csv(projected_test_data, paste0(dirname,"/test_",b,".csv"))
